### Detailed exploration of TCGA subtypes

In the "TCGA Cencer Types" tab, you can dive deeper into specific projects within the TCGA Tumor dataset. Select a TCGA project and further analyze the data by grouping scatter plots based on experimental variables such as gender, mRNA subtype, or DNA methylation subtype. Similar to the first tab, you'll set the model parametersâ€”deciding whether to use raw, ranged, or scaled RAS values and selecting the desired RAS adjustment method.

Beneath the initial GD2 score visualization, you can conduct differential gene expression analysis (DEA) for the entire project or a specific subset, like mRNA subtype LGr4 in the Glioblastoma Multiforme project. Use the GD2 score for stratification, with options to divide samples into GD2-high and GD2-low groups using the median, set a specific GD2 threshold, or create three groups based on percentiles. Customize DEA settings, including False Discovery Rate, filtering, and hypothesis weighting. The results display overall gene regulation, a searchable table of significant genes, and additional details when a gene is selected. The tab also includes diagnostic and summary plots, such as MA and interactive volcano plots. You get an overall result of the DEA describing, how many genes were identified as up or down regulated. The "DEA Genes" table allows you to see and search for genes and their statistics. Select a gene by clicking on the row in the table and see further information on the gene as well as the distribution between the GD2-high and low group in the next two boxes. The two diagnostic plots provide further information on the DEA result, as well as the two summary plots - the MA plot and the interactive volcano plot. 