<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>GD2Viz app - Explore public datasets and analyze your data • GD2Viz</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GD2Viz app - Explore public datasets and analyze your data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GD2Viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.01.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/GD2Viz-app.html">GD2Viz app - Explore public datasets and analyze your data</a></li>
    <li><a class="dropdown-item" href="../articles/Predict-GD2-Score.html">Computing GD2 Scores Programmatically Using GD2Viz</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/arsenij-ust/GD2Viz/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="GD2Viz-app_files/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="GD2Viz-app_files/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>GD2Viz app - Explore public datasets and analyze your data</h1>
                        <h4 data-toc-skip class="author">Arsenij
Ustjanzew</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:arsenij.ustjanzew@gmail.com" class="email">arsenij.ustjanzew@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">21 Mai 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/arsenij-ust/GD2Viz/blob/HEAD/vignettes/GD2Viz-app.Rmd" class="external-link"><code>vignettes/GD2Viz-app.Rmd</code></a></small>
      <div class="d-none name"><code>GD2Viz-app.Rmd</code></div>
    </div>

    
    
<p><br><strong>Authors</strong>: Arsenij Ustjanzew [aut, cre] (<a href="https://orcid.org/0000-0000-0000-0000" class="external-link uri">https://orcid.org/0000-0000-0000-0000</a>), Claudia Paret
[aut] (<a href="https://orcid.org/0000-0000-0000-0000" class="external-link uri">https://orcid.org/0000-0000-0000-0000</a>), Federico Marini
[aut] (<a href="https://orcid.org/0000-0003-3252-7758" class="external-link uri">https://orcid.org/0000-0003-3252-7758</a>)<br><strong>Version</strong>: 0.01.1<br><strong>Compiled date</strong>:
2025-05-21<br><strong>License</strong>: MIT + file LICENSE<br></p>
<p><img src="images/GD2Viz6.png" align="center" width="150px"></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/arsenij-ust/GD2Viz" class="external-link">GD2Viz</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction-and-scope">Introduction and scope<a class="anchor" aria-label="anchor" href="#introduction-and-scope"></a>
</h2>
<p>GD2Viz is a specialized visualization and analysis tool designed to
explore the tumor-specific expression of GD2 — a disialoganglioside of
growing importance in cancer immunotherapy. GD2 has emerged as a
clinically actionable biomarker, due to its restricted expression in
normal tissues and its elevated presence in malignancies such as
neuroblastoma. Accurately assessing GD2 expression across tumor entities
is pivotal for identifying patients who may benefit from targeted
therapies, particularly anti-GD2 immunotherapeutics.</p>
<p>Building on the methodology proposed by Ustjanzew et al. (2024),
GD2Viz integrates transcriptomic data with a curated model of
glycosphingolipid metabolism to compute and interpret GD2 Scores. These
scores reflect the net activity of pathway-level reactions called
Reaction Activity Score (RAS) classified as GD2-promoting or
GD2-mitigating. The framework is implemented as both an R package and an
interactive Shiny application, empowering users to explore precomputed
scores in large-scale public datasets or to analyze their own custom
RNA-Seq data.</p>
<p>This vignette focuses on the Shiny application component of GD2Viz,
detailing its core features and guiding users through its analytical
capabilities.</p>
<p>While GD2Viz provides a powerful lens for examining the
glycosphingolipid metabolism as it relates to GD2, users are advised to
interpret GD2 Scores within the context of their computational origin.
These scores are inferential by nature and should complement, not
replace, direct experimental validation. Nevertheless, GD2Viz offers a
scalable, transparent, and biologically grounded platform to advance the
understanding and clinical utility of GD2 expression profiles in cancer
research.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>GD2Viz is a stand-alone R package, so a user can start and use the
application locally. However, the local version of GD2Viz does not
include the exploration of public RNA-seq datasets and allows you only
to load your own data tocompute the sphingolipid-pathway RAS scoresand
predict the GD2 score for your samples. Please use the online version at
<a href="http://shiny.imbei.uni-mainz.de:3838/GD2Viz" class="external-link uri">http://shiny.imbei.uni-mainz.de:3838/GD2Viz</a> to access
the extended version and explore the GD2 scores of available public
datasets.</p>
<p>A local installation of the latest version of R and RStudio is
required.</p>
<p>The package can be installed with the <em><a href="https://CRAN.R-project.org/package=remotes" class="external-link">remotes</a></em>
library:</p>
<hr>
<p>When using GD2Viz for your research, please cite: Ustjanzew et
al. Predicting GD2 expression across cancer types by the integration of
pathway topology and transcriptome data. 2025</p>
<hr>
<div class="section level3">
<h3 id="installation-1">Installation<a class="anchor" aria-label="anchor" href="#installation-1"></a>
</h3>
<p><strong>GD2Viz</strong> is available as a standalone R package,
enabling users to run the application locally. The local version
supports the analysis of custom RNA-Seq datasets by computing Reaction
Activity Scores (RAS) for the glycosphingolipid pathway and predicting
GD2 Scores for user-provided samples.</p>
<blockquote>
<p>⚠️ <strong>Note:</strong> The local version <strong>does not</strong>
include access to preloaded public datasets (e.g., TCGA, GTEx, TARGET).
To explore GD2 Scores across these datasets, please use the online
version of the app, available at: <a href="http://shiny.imbei.uni-mainz.de:3838/GD2Viz" class="external-link">http://shiny.imbei.uni-mainz.de:3838/GD2Viz</a></p>
</blockquote>
<div class="section level4">
<h4 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h4>
<p>Ensure that the latest versions of <strong>R</strong> and
<strong>RStudio</strong> are installed on your system.</p>
</div>
<div class="section level4">
<h4 id="installation-via-remotes">Installation via <code>remotes</code><a class="anchor" aria-label="anchor" href="#installation-via-remotes"></a>
</h4>
<p>You can install GD2Viz using the <a href="https://CRAN.R-project.org/package=remotes" class="external-link"><code>remotes</code></a>
package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"arsenij-ust/GD2Viz"</span><span class="op">)</span></span></code></pre></div>
<p>Once installed, you can launch the application locally by
running:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">GD2Viz</span><span class="fu">::</span><span class="fu"><a href="../reference/GD2Viz.html">GD2Viz</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>For advanced usage, including programmatic access to functions for
GD2 Score calculation and pathway analysis, refer to the vignette
<todo>.</todo></p>
<p>The application is organized into three main tabs, each of which is
described in detail in the following sections:</p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="tab-public-datasets-online-version">Tab ‘Public Datasets’ (online version)<a class="anchor" aria-label="anchor" href="#tab-public-datasets-online-version"></a>
</h2>
<p><img src="images/tab1.png" align="center" width="80%"></p>
<p>The <strong>Public Datasets</strong> tab provides an interactive
interface to explore GD2 Scores across six major RNA-Seq datasets:</p>
<ul>
<li><strong>TCGA Tumor</strong></li>
<li><strong>TCGA Normal</strong></li>
<li><strong>GTEx</strong></li>
<li><strong>TARGET</strong></li>
<li><strong>St. Jude Cloud</strong></li>
<li>
<strong>CBTTC</strong> (from the Pediatric Brain Tumor Atlas)</li>
</ul>
<p>The RAS matrices were precomputed and are used to train a Support
Vector Model (SVM) with linear kernel on the fly. The</p>
<p>To ensure rapid and reproducible analysis, both adjusted and
unadjusted <strong>Reaction Activity Scores (RAS)</strong> have been
precomputed for all public datasets. Additionally, the RAS matrices of
the training dataset serve as input for a <strong>Support Vector Machine
(SVM)</strong> with a linear kernel, which is trained on-the-fly using
the specified preprocessing and adjustment settings. The trained model
is then used to predict GD2 Scores across the selected dataset.</p>
<p>Users can control two key preprocessing settings:</p>
<div class="section level4">
<h4 id="ras-processing-method">RAS Processing Method<a class="anchor" aria-label="anchor" href="#ras-processing-method"></a>
</h4>
<p>Determines how the RAS values are prepared before being used for
model training:</p>
<ul>
<li>
<strong>Raw</strong>: Uses unmodified RAS values as computed from
gene expression.</li>
<li>
<strong>Ranged</strong>: Normalizes each reaction’s activity score
to a [0, 1] range across all samples.</li>
<li>
<strong>Scaled</strong>: Standardizes RAS values to have zero mean
and unit variance (z-score transformation).</li>
</ul>
</div>
<div class="section level4">
<h4 id="ras-adjustment-type">RAS Adjustment Type<a class="anchor" aria-label="anchor" href="#ras-adjustment-type"></a>
</h4>
<p>Specifies how the RAS values are adjusted to reflect network topology
and biological dependencies:</p>
<ul>
<li>
<strong>Unadjusted RAS</strong>: Uses the direct output of the RAS
calculation, based purely on transcriptomic input.</li>
<li>
<strong>Adjusted by Transition Probability</strong>: Weighs RAS
values according to pathway transition probabilities, incorporating
pathway structure.</li>
<li>
<strong>Recursive Adjustment</strong>: Further refines RAS values
through iterative propagation along the reaction network, capturing
upstream and downstream pathway effects.</li>
</ul>
<p>These user-defined parameters directly influence how the training
data is interpreted by the SVM model and therefore affect the resulting
GD2 Score predictions. This design allows users to explore how different
biological assumptions and preprocessing choices impact the
characterization of GD2 expression across diverse RNA-Seq datasets.</p>
</div>
<div class="section level4">
<h4 id="visualization-options">Visualization Options<a class="anchor" aria-label="anchor" href="#visualization-options"></a>
</h4>
<p>The GD2 Score can be visualized using three plot types:</p>
<ul>
<li>
<strong>Scatter plot</strong>: Ideal for viewing individual sample
distributions</li>
<li>
<strong>Box plot</strong>: Useful for summarizing group-level
differences</li>
<li>
<strong>Violin plot</strong>: Highlights both summary statistics and
distribution density</li>
</ul>
<p>Users can also normalize GD2 Scores to a [0–1] range for
cross-dataset comparison.</p>
</div>
<div class="section level4">
<h4 id="grouping-and-highlighting">Grouping and Highlighting<a class="anchor" aria-label="anchor" href="#grouping-and-highlighting"></a>
</h4>
<p>Samples can be grouped by metadata variables (e.g.,
<em>Primary_Disease</em>) to contextualize GD2 Score distributions
across conditions. Additionally, users may highlight specific groups —
for example, highlighting <strong>Glioblastoma Multiforme</strong>
within the TCGA dataset.</p>
<p>This tab is ideal for hypothesis generation, comparative analysis of
cancer types with elevated GD2 expression that may be candidates for
targeted anti-GD2 therapy.</p>
</div>
<div class="section level4">
<h4 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h4>
<p>The zero value of the GD2 Score represents the SVM hyper plane
dividing GD2-high samples from GD2-low samples. The GD2 Score of a
sample is equal to the SVM decision weight of the sample. This means,
samples with GD2 Score far below zero are GD2 negative, samples close to
zero have an unclear/intermediate GD2 expression and samples far above
zero might express high levels of GD2.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="tab-tcga-cancer-types-online-version">Tab ‘TCGA Cancer Types’ (online version)<a class="anchor" aria-label="anchor" href="#tab-tcga-cancer-types-online-version"></a>
</h2>
<p><img src="images/tab2_1.png" align="center" width="80%"></p>
<p>The <strong>TCGA Cancer Types</strong> tab enables detailed
exploration of individual tumor types within the TCGA dataset. Users can
select a specific TCGA project (e.g., <em>Glioblastoma Multiforme</em>)
and visualize GD2 Scores across clinically or molecularly defined
subgroups, such as mRNA subtypes, DNA methylation classes, or other
available metadata variables.</p>
<p>As in the <strong>Public Datasets</strong> tab, users can configure
key model parameters that influence the GD2 Score prediction:</p>
<ul>
<li>
<strong>RAS processing method</strong>: raw, ranged, or scaled</li>
<li>
<strong>RAS adjustment type</strong>: unadjusted, adjusted by
transition probability, or recursively adjusted</li>
</ul>
<p>The resulting plot displays GD2 Scores grouped by the selected
stratification variable, providing insight into intratumoral
heterogeneity and subtype-specific GD2 expression.</p>
<hr>
<div class="section level4">
<h4 id="differential-expression-analysis-dea">Differential Expression Analysis (DEA)<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-dea"></a>
</h4>
<p><img src="images/tab2_2.png" align="center" width="80%"></p>
<p>Beyond visualization, this tab includes a powerful
<strong>Differential Expression Analysis</strong> module to investigate
transcriptomic differences between GD2-defined sample groups. Users can
stratify samples based on GD2 Scores using:</p>
<ul>
<li>A median split (GD2-high vs. GD2-low)</li>
<li>A custom threshold value</li>
<li>Upper- vs. lower GD2 Score percentiles</li>
</ul>
<p>Additionally, the user can select between DESeq2 and limma-voom
pipelines to compute the DEA. If you want to analyze more than 20
samples, we suggest limma-voom because the computation will be
faster.</p>
</div>
<div class="section level4">
<h4 id="dea-output">DEA Output<a class="anchor" aria-label="anchor" href="#dea-output"></a>
</h4>
<p>The DEA results section offers a multi-faceted view of differential
gene regulation:</p>
<ul>
<li>
<strong>Summary statistics</strong>: Total number of upregulated and
downregulated genes</li>
<li>
<strong>Searchable results table</strong>: View, filter, and sort
significant genes</li>
<li>
<strong>Gene-specific details</strong>: Select any gene from the
table to see additional annotation and visualize its expression
distribution across GD2-defined groups</li>
<li>
<strong>Interactive volcano plot</strong>: Visualize effect size and
significance</li>
<li>
<strong>MA plot</strong>: Assess expression fold changes versus mean
expression</li>
<li>
<strong>P-value Histogram</strong>: Assesses the distribution of raw
p-values to evaluate test validity and potential signal</li>
<li>
<strong>Log2 Fold-Change Histogram</strong>: Provides a global
overview of the magnitude and directionality of differential gene
expression</li>
</ul>
<p>This integrated analysis allows users to link GD2 Score variation
with broader transcriptomic changes, helping to generate hypotheses
about GD2-associated pathways, biomarkers, or therapeutic targets within
specific cancer types.</p>
<p>The results of the DEA can be downloaded.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="tab-analyze-your-data-local-and-online-versions">Tab ‘Analyze Your Data’ (local and online versions)<a class="anchor" aria-label="anchor" href="#tab-analyze-your-data-local-and-online-versions"></a>
</h2>
<p><img src="images/tab3_1.png" align="center" width="80%"></p>
<p>The <strong>Analyze Your Data</strong> tab enables users to upload
their own <strong>RNA-Seq datasets</strong> and compute both
<strong>Reaction Activity Scores (RAS)</strong> and <strong>GD2
Scores</strong> using the GD2Viz framework. This tab provides a full
analytical pipeline, from raw counts to visualization and
interpretation, all within a user-friendly interface.</p>
<div class="section level4">
<h4 id="step-1-data-upload">Step 1: Data Upload<a class="anchor" aria-label="anchor" href="#step-1-data-upload"></a>
</h4>
<p>Users can upload their data in one of two formats:</p>
<ul>
<li>
<p><strong>Count matrix and metadata</strong> (<code>.tsv</code>
format):</p>
<ul>
<li>The count matrix should contain <strong>raw, unnormalized
counts</strong>, with rows labeled by <strong>Gene Symbols</strong> and
columns representing samples.</li>
<li>The metadata file should contain sample-level annotations. Column
names in the count matrix must match the row names in the metadata
file.</li>
</ul>
</li>
<li>
<p><strong>DESeqDataSet object</strong> (<code>.rds</code>
format):</p>
<ul>
<li>This is a serialized R object containing counts and metadata,
typically generated from the DESeq2 pipeline.</li>
</ul>
</li>
</ul>
<p>Text file with count data:</p>
<p><img src="images/tab3_7.png" align="center" width="90%"></p>
<p>Text file with metadata (exerimental variables):</p>
<p><img src="images/tab3_8.png" align="center" width="15%"></p>
<blockquote>
<p>⚠️ Make sure your input files are properly formatted - same sample
names and order. The metadata file can include experimental variables
such as subtype, treatment, or batch for later grouping and
analysis.</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="step-2-ras-computation">Step 2: RAS Computation<a class="anchor" aria-label="anchor" href="#step-2-ras-computation"></a>
</h4>
<p>After successful data upload, click <strong>“Compute Reaction
Activity Scores”</strong> to calculate the raw and adjusted RAS
matrices. These values represent the metabolic activity of reactions in
the glycosphingolipid pathway.</p>
</div>
<div class="section level4">
<h4 id="step-3-gd2-score-prediction">Step 3: GD2 Score Prediction<a class="anchor" aria-label="anchor" href="#step-3-gd2-score-prediction"></a>
</h4>
<p>Users can select how RAS values should be preprocessed for model
training:</p>
<ul>
<li>
<strong>Raw</strong>: Unmodified RAS values</li>
<li>
<strong>Ranged</strong>: Normalized to [0, 1] range</li>
<li>
<strong>Scaled</strong>: Standardized to zero mean and unit
variance</li>
</ul>
<p>Once selected, click <strong>“Compute GD2 Score”</strong> to train
the SVM model and predict GD2 Scores based on each RAS matrix variant.
The outputs can be downloaded directly as <code>.tsv</code> files.</p>
</div>
<div class="section level4">
<h4 id="step-4-visualization-and-interpretation">Step 4: Visualization and Interpretation<a class="anchor" aria-label="anchor" href="#step-4-visualization-and-interpretation"></a>
</h4>
<p>The <strong>Global Plot Settings</strong> panel lets you choose which
adjusted RAS type to visualize and allows you to <strong>group samples
by any experimental variable</strong> defined in your metadata.</p>
<hr>
</div>
<div class="section level4">
<h4 id="visualization-panels">Visualization Panels:<a class="anchor" aria-label="anchor" href="#visualization-panels"></a>
</h4>
<ol style="list-style-type: decimal">
<li>
<p><strong>RAS Heatmap</strong></p>
<ul>
<li>Displays the computed RAS matrix as an adjustable heatmap.</li>
<li>Customize row/column labels, scaling, clustering method, and
distance metrics by clicking on the gears symbol.</li>
</ul>
</li>
</ol>
<p><img src="images/tab3_4.png" align="center" width="60%"></p>
<ol start="2" style="list-style-type: decimal">
<li>
<p><strong>Promoting vs. Diminishing Reactions Scatter
Plot</strong></p>
<ul>
<li>Visualizes the sum of GD2-promoting vs. GD2-diminishing reactions
used by the SVM model as predictive variables. The lines represent the
GD2 Score scale, where zero is the hyper plane of the SVM model.</li>
</ul>
</li>
</ol>
<p><img src="images/tab3_2.png" align="center" width="60%"></p>
<ol start="3" style="list-style-type: decimal">
<li>
<p><strong>Predicted GD2 Score Plot</strong></p>
<ul>
<li>Visualize GD2 scores using scatter, box, or violin plots.</li>
<li>When scatter is selected, you can optionally plot GD2 scores against
a selected gene of interest.</li>
</ul>
</li>
</ol>
<p>The zero value of the GD2 Score represents the SVM hyper plane
dividing GD2-high samples from GD2-low samples. The GD2 Score of a
sample is equal to the SVM decision weight of the sample. This means,
samples with GD2 Score far below zero are GD2 negative, samples close to
zero have an unclear/intermediate GD2 expression and samples far above
zero might express high levels of GD2.</p>
<p><img src="images/tab3_6.png" align="center" width="60%"></p>
<ol start="4" style="list-style-type: decimal">
<li>
<p><strong>GD2 Score vs. Stemness Score</strong></p>
<ul>
<li>Explores the relationship between GD2 expression and cellular
stemness, which may inform interpretation of differentiation
states.</li>
</ul>
</li>
</ol>
<p>Refer for the Stemness score: Malta TM, Sokolov A, et al.; Cancer
Genome Atlas Research Network; Stuart JM, Hoadley KA, Laird PW,
Noushmehr H, Wiznerowicz M. Machine Learning Identifies Stemness
Features Associated with Oncogenic Dedifferentiation. Cell. 2018 Apr
5;173(2):338-354.e15. doi: 10.1016/j.cell.2018.03.034.</p>
<p><img src="images/tab3_3.png" align="center" width="30%"></p>
<hr>
</div>
<div class="section level4">
<h4 id="group-comparison-and-pathway-analysis">Group Comparison and Pathway Analysis<a class="anchor" aria-label="anchor" href="#group-comparison-and-pathway-analysis"></a>
</h4>
<p><img src="images/tab3_5.png" align="center" width="100%"></p>
<p>Users can compare two experimental groups to examine shifts in
glycosphingolipid metabolism:</p>
<ul>
<li><p><strong>Log2 Fold Change Plot</strong>: Highlights up- and
downregulated reactions within the glycosphingolipid pathway.</p></li>
<li><p><strong>Ganglioside Pathway Visualization</strong>: Provides a
detailed graphical analysis of ganglioside-specific reactions.</p></li>
</ul>
<hr>
</div>
<div class="section level4">
<h4 id="differential-expression-analysis-dea-1">Differential Expression Analysis (DEA)<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-dea-1"></a>
</h4>
<p>The final section mirrors the functionality of the <strong>TCGA
Cancer Types</strong> tab. Users can perform DEA using GD2 score-based
stratification:</p>
<ul>
<li><p>Define GD2-high vs. GD2-low groups using the median, percentiles,
or custom thresholds</p></li>
<li>
<p>Run the analysis and visualize results with:</p>
<ul>
<li>
<strong>Summary stats</strong> on up/downregulated genes</li>
<li><strong>Interactive volcano plot</strong></li>
<li><strong>MA plot</strong></li>
<li><strong>P-value histogram</strong></li>
<li><strong>Log2 fold-change histogram</strong></li>
<li><strong>Searchable results table</strong></li>
<li><strong>Gene-specific expression plots</strong></li>
</ul>
</li>
</ul>
<p>This allows comprehensive exploration of transcriptomic changes
associated with GD2 phenotypes in user-defined datasets.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arsenij Ustjanzew, Claudia Paret, Federico Marini.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
