<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute Reaction Activity Scores for Custom Dataset — computeReactionActivityScores • GD2Viz</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute Reaction Activity Scores for Custom Dataset — computeReactionActivityScores"><meta name="description" content="The computeReactionActivityScores function computes reaction activity scores for a custom dataset using the provided metabolic network, training data geometric means, and other parameters. It integrates gene expression data with the metabolic network to derive reaction activity scores, transition probabilities, and adjusted reaction activity matrices."><meta property="og:description" content="The computeReactionActivityScores function computes reaction activity scores for a custom dataset using the provided metabolic network, training data geometric means, and other parameters. It integrates gene expression data with the metabolic network to derive reaction activity scores, transition probabilities, and adjusted reaction activity matrices."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GD2Viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.01.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/GD2Viz-app.html">GD2Viz app - Explore public datasets and analyze your data</a></li>
    <li><a class="dropdown-item" href="../articles/Predict-GD2-Score.html">Computing GD2 Scores Programmatically Using GD2Viz</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/arsenij-ust/GD2Viz/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Compute Reaction Activity Scores for Custom Dataset</h1>
      <small class="dont-index">Source: <a href="https://github.com/arsenij-ust/GD2Viz/blob/HEAD/R/utils.R" class="external-link"><code>R/utils.R</code></a></small>
      <div class="d-none name"><code>computeReactionActivityScores.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The <code>computeReactionActivityScores</code> function computes reaction activity scores for a custom dataset using the provided metabolic network, training data geometric means, and other parameters. It integrates gene expression data with the metabolic network to derive reaction activity scores, transition probabilities, and adjusted reaction activity matrices.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">computeReactionActivityScores</span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dds <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">mgraph</span>,</span>
<span>  <span class="va">geom</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>A data frame or matrix containing raw unnormalized RNA-Seq gene counts. Columns are samples, rows are genes.</p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>A data frame containing sample metadata. Required if counts is provided.</p></dd>


<dt id="arg-dds">dds<a class="anchor" aria-label="anchor" href="#arg-dds"></a></dt>
<dd><p>A DESeqDataSet object. If provided, counts and metadata are ignored.</p></dd>


<dt id="arg-mgraph">mgraph<a class="anchor" aria-label="anchor" href="#arg-mgraph"></a></dt>
<dd><p>The graph object representing the metabolic network.</p></dd>


<dt id="arg-geom">geom<a class="anchor" aria-label="anchor" href="#arg-geom"></a></dt>
<dd><p>A named numeric vector containing the geometric means of the training dataset.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing various Reaction Activity scores computed for the custom dataset based on the provided metabolic network and training data.</p>
<p>custom_coldata The modified sample metadata corresponding to the custom dataset.</p>
<p>ras The original Reaction Activity matrix computed for the custom dataset.</p>
<p>ras_prob The Reaction Activity matrix adjusted by transition probabilities.</p>
<p>ras_prob_path The Reaction Activity matrix adjusted by transition probabilities considering a target node.</p>
<p>ras_prob_rec The Reaction Activity matrix adjusted by transition probabilities with a recursive adjustment method.</p>
<p>ras_prob_up The Reaction Activity matrix adjusted by adding transition probabilities.</p>
<p>ras_prob_up_path The Reaction Activity matrix adjusted by adding transition probabilities considering a target node.</p>
<p>ras_prob_up_rec The Reaction Activity matrix adjusted by adding transition probabilities with a recursive adjustment method.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function integrates gene expression data with a metabolic network to compute reaction activity scores. It estimates size factors for the custom dataset based on the geometric means of the training dataset and normalizes the gene expression data accordingly. The function then computes Reaction Activity scores and transition probabilities for the custom dataset using the provided metabolic network. Various adjustments are applied to the Reaction Activity scores to derive different metrics, including those considering transition probabilities, target nodes, and recursive adjustment methods.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code>DESeqDataSetFromMatrix</code>, <code>estimateSizeFactorsForMatrix</code>, <code><a href="computeReactionActivity.html">computeReactionActivity</a></code>, <code><a href="computeTransitionProbability.html">computeTransitionProbability</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Load necessary data</span></span></span>
<span class="r-in"><span><span class="va">geom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./SVM_GD2_dashboard/data/geom_train_data.Rds"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">custom_counts</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"SVM_GD2_dashboard/data/datasets/custom_counts.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">custom_coldata</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"SVM_GD2_dashboard/data/datasets/custom_metadata.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mgraph</span> <span class="op">&lt;-</span> <span class="fu">load_metabolic_network</span><span class="op">(</span><span class="st">"path/to/metabolic_network.graphml"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compute Reaction Activity scores for the custom dataset</span></span></span>
<span class="r-in"><span><span class="va">reaction_activity</span> <span class="op">&lt;-</span> <span class="fu">computeReactionActivityScores</span><span class="op">(</span><span class="va">custom_counts</span>, <span class="va">custom_coldata</span>, <span class="va">mgraph</span>, <span class="va">geom</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arsenij Ustjanzew, Claudia Paret, Federico Marini.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

